File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        
        def min_moves(node):
            nonlocal ans
            if(not node):
                return 0
            
            left = min_moves(node.left)
            right = min_moves(node.right)
            ret = node.val + left + right - 1
            # ex = (node.val + e1 + e2 - 1)
            # ne = abs(ex - n1 + n2)
            ans += abs(ret)
            return ret
        
        
        
        ans = 0
        min_moves(root)
        return ans

File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func distributeCoins(root *TreeNode) int {
    var ops = 0
    disCoins(root, &ops)
    return ops
}

func disCoins(root *TreeNode, ops *int) int {
    if root == nil {
        return 0
    }
    
    var left = disCoins(root.Left, ops)
    var right = disCoins(root.Right, ops)
    // fmt.Println(left, right, root.Val, root.Val + left + right - 1)
    *ops += int(math.Abs(float64(left))) + int(math.Abs(float64(right))) 
    return root.Val + left + right - 1
}